File: src/api/tauri/drivers.ts
/**
 * Typed Tauri IPC wrappers for Driver Intelligence commands.
 *
 * These are the ONLY entry points for Tauri `invoke` calls related to
 * drivers. Components and pages MUST NOT call `invoke` directly
 * (doc 48 AC-FS-04).
 *
 * Command names are snake_case (Tauri convention).
 */

import { invoke } from '@tauri-apps/api/core';
import type {
  DriverInfo,
  DriverUpdateInstallResult,
  DriverUpdateScanResult,
  GpuCleanPlan,
  GpuCleanResult,
  VaultEntry,
} from '../../types/device.types';

/** Scan installed drivers, score health, and persist results. */
export const scanDrivers = (): Promise<DriverInfo[]> =>
  invoke<DriverInfo[]>('scan_drivers');

/** List previously scanned drivers for the local device. */
export const listDrivers = (): Promise<DriverInfo[]> =>
  invoke<DriverInfo[]>('list_drivers');

/** Search Windows Update for available driver packages. */
export const scanDriverUpdates = (): Promise<DriverUpdateScanResult> =>
  invoke<DriverUpdateScanResult>('scan_driver_updates');

/** Download + install selected Windows Update driver packages. */
export const installDriverUpdates = (
  updateIds: string[],
  confirm: boolean,
): Promise<DriverUpdateInstallResult> =>
  invoke<DriverUpdateInstallResult>('install_driver_updates', {
    updateIds,
    confirm,
  });

/** Dry-run DDU-class GPU driver clean plan. */
export const previewGpuDriverClean = (
  vendor?: string | null,
): Promise<GpuCleanPlan> =>
  invoke<GpuCleanPlan>('preview_gpu_driver_clean', {
    vendor: vendor ?? null,
  });

/** System Restore point for the GPU clean gate. */
export const createGpuCleanRestorePoint = (): Promise<VaultEntry> =>
  invoke<VaultEntry>('create_gpu_clean_restore_point');

/** Execute allowlisted package removal after restore point + confirm. */
export const executeGpuDriverClean = (opts: {
  planId: string;
  restorePointId: string;
  vendor?: string | null;
  confirm: boolean;
}): Promise<GpuCleanResult> =>
  invoke<GpuCleanResult>('execute_gpu_driver_clean', {
    planId: opts.planId,
    restorePointId: opts.restorePointId,
    vendor: opts.vendor ?? null,
    confirm: opts.confirm,
  });

File: src-tauri/src/commands/drivers.rs
//! Driver Intelligence IPC.

use tauri::State;

use crate::drivers;
use crate::error::CoreError;
use crate::models::{
    DriverInfo, DriverUpdateInstallResult, DriverUpdateScanResult, GpuCleanPlan, GpuCleanResult,
    VaultEntry,
};
use crate::storage::driver_repo;
use crate::AppState;

#[tauri::command]
pub fn scan_drivers(state: State<'_, AppState>) -> Result<Vec<DriverInfo>, CoreError> {
    let conn = state.conn()?;
    drivers::scan_drivers(&conn)
}

#[tauri::command]
pub fn list_drivers(state: State<'_, AppState>) -> Result<Vec<DriverInfo>, CoreError> {
    let conn = state.conn()?;
    driver_repo::list_drivers(&conn)
}

/// Search Windows Update for available driver packages (Type=Driver).
#[tauri::command]
pub fn scan_driver_updates(
    state: State<'_, AppState>,
) -> Result<DriverUpdateScanResult, CoreError> {
    let conn = state.conn()?;
    drivers::scan_driver_updates(&conn)
}

/// Download + install selected Windows Update driver packages.
#[tauri::command]
pub fn install_driver_updates(
    state: State<'_, AppState>,
    update_ids: Vec<String>,
    confirm: bool,
) -> Result<DriverUpdateInstallResult, CoreError> {
    let conn = state.conn()?;
    drivers::install_driver_updates(&conn, update_ids, confirm)
}

/// Dry-run GPU driver clean plan (DDU-class). `vendor`: nvidia|amd|intel|auto.
#[tauri::command]
pub fn preview_gpu_driver_clean(
    state: State<'_, AppState>,
    vendor: Option<String>,
) -> Result<GpuCleanPlan, CoreError> {
    let conn = state.conn()?;
    drivers::preview_gpu_driver_clean(&conn, vendor)
}

/// Create a System Restore point specifically for the GPU clean wizard gate.
#[tauri::command]
pub fn create_gpu_clean_restore_point(state: State<'_, AppState>) -> Result<VaultEntry, CoreError> {
    let conn = state.conn()?;
    drivers::create_gpu_clean_restore_point(&conn)
}

/// Execute allowlisted pnputil package removal after restore point + confirm.
#[tauri::command]
pub fn execute_gpu_driver_clean(
    state: State<'_, AppState>,
    plan_id: String,
    restore_point_id: String,
    vendor: Option<String>,
    confirm: bool,
) -> Result<GpuCleanResult, CoreError> {
    let conn = state.conn()?;
    drivers::execute_gpu_driver_clean(&conn, plan_id, restore_point_id, vendor, confirm)
}
